coding/pybind11_style_guide.rst

```diff
@@ -0,0 +1,592 @@
+
+By wrapping different headers into separate modules (to be combined in ``__init__``) we make builds more parallelize able, make it easier to avoid circular dependencies, and make partial rebuilds faster.
+
+If a group of headers together provide functionality that cannot be used independently, they may be wrapped into a single module.  The headers wrapped by such a module must be prominently listed in a comment near the top of the source file.
+
+.. _style-guide-pybind11-subpackage:
+
+Wrappers that contain both C++ and Python code MUST define a subpackage
+-----------------------------------------------------------------------
+
+When the wrappers for a header (or group of closely-related headers) require both C++ and Python, both files MUST be moved to a new Python subpackage, with an ``__init__.py`` file that lifts all public symbols from both modules to package scope.  The Python module need not export symbols also provided by the C++ module (frequently, it will simply modify them, by e.g. adding methods to classes using the ``lsst.utils.continueClass`` decorator).  The C++ module name should still be the lowercased header file name, and the Python module name MUST be this with a "Continued" suffix.
+
+For example, for header file ``LinearTransform.``, we would have::
+
+    linearTransform/linearTransform.cc:
+        <C++ wrappers>
+
+    linearTransform/linearTransformContinued.py:
+        <Python extensions to the wrappers>
+
+    linearTransform/__init__.py:
+        from .linearTransform import *
+        from .linearTransformContinued import *
+
+.. _style-guide-pybind11-cpp-vs-python:
+
+Trivial extensions to wrappers SHOULD be implemented in C++
+-----------------------------------------------------------
+
+Simple extensions such as ``__repr__`` or ``__reduce__`` should be implemented via lambdas in compiled modules, utilizing the pybind11 Python C++ API (e.g. ``pybind11::object``) as necessary.
+
+Longer extensions that involve significant logic or language constructs difficult to implement using the C++ Python API (e.g. generators) should go in pure-Python files.
+
+This rule applies regardless of whether a pure-Python extension module already exists; this prevents the correct code organization from becoming a function of history.
+
+Using pure-Python modules only when necessary minimizes the number of source files and helps keep class definitions together.
+
+.. _style-guide-pybind11-include:
+
+Pybind11 headers should precede all other headers in the include ordering
+-------------------------------------------------------------------------
+
+``pybind11.h`` includes ``Python.h`` and must hence be included before all other headers.
+To keep a reasonable grouping, all other pybind11 headers should be included in this same include block.
```

Comment:
Should link to https://docs.python.org/3/c-api/intro.html#include-files for explanation.